nexus/src/db/error.rs

```diff
@@ -80,64 +80,52 @@ pub fn diesel_pool_result_optional<T>(
     }
 }
 
-/// Converts a Diesel pool error to an external error when operating on a
-/// particular resource, identified by resource type and lookup type
-/// NOTE: If you already have an `authz::ApiResource`, you should use
-/// [`public_error_from_diesel_pool_authz()`] instead.  Eventually, the only
-/// uses of this function should be in the DataStore functions that actually
-/// look up a record for the first time.
-pub fn public_error_from_diesel_pool_lookup(
-    error: PoolError,
-    resource_type: ResourceType,
-    lookup_type: LookupType,
-) -> PublicError {
-    public_error_from_diesel_pool_helper(error, |error| {
-        public_error_from_diesel_lookup_helper(error, || {
-            PublicError::ObjectNotFound {
-                type_name: resource_type,
-                lookup_type,
-            }
-        })
-    })
-}
-
-/// Converts a Diesel pool error to an external error when operating on a
-/// particular resource, identified by ApiResource
-pub fn public_error_from_diesel_pool_authz<R>(
-    error: PoolError,
-    resource: &R,
-) -> PublicError
-where
-    R: crate::authz::ApiResource,
-{
-    public_error_from_diesel_pool_helper(error, |error| {
-        public_error_from_diesel_lookup_helper(error, || resource.not_found())
-    })
-}
-
-/// Converts a Diesel pool error to an external error,
-/// when requested as part of a creation operation
-pub fn public_error_from_diesel_pool_create(
-    error: PoolError,
-    resource_type: ResourceType,
-    object_name: &str,
-) -> PublicError {
-    public_error_from_diesel_pool_helper(error, |error| {
-        public_error_from_diesel_create(error, resource_type, object_name)
-    })
+/// Describes the operation type which triggered the error,
+/// as well as context which may be used to populate more informative errors.
+pub enum OpKind<'a> {
+    /// The operation was performed on a particular [`ApiResource`].
+    ///
+    /// NOTE: Practically, all objects which implement [`ApiResourceError`]
+    /// also implement [`ApiResource`]. However, [`ApiResource`] is not object
+    /// safe because it implements [`std::clone::Clone`].
+    Authz(&'a dyn crate::authz::ApiResourceError),
+    /// The operation was attempting to lookup or update a resource.
+    Lookup(ResourceType, LookupType),
+    /// The operation was attempting to create a resource with a name.
+    Create(ResourceType, &'a str),
+    /// All other operation types.
+    Other,
```

Comment:
I guess I could call this `OpKind::Infallible`, but that still seems misleading - it *can* fail, in the same way the `public_error_from_diesel_pool_shouldnt_fail` could. I think part of the weirdness comes from the other variants being phrased in terms of the input operation kind (lookup, create) and this one being phrased in terms of the expected error types. If we still think of this case in terms of input operation, it really is a catch-all for "something that isn't all the other cases", and if we don't want that to return exclusively 500 errors, a different variant should be added.

I'll at least update the comment indicating the expected usage.

Reply:
> I think part of the weirdness comes from the other variants being phrased in terms of the input operation kind (lookup, create) and this one being phrased in terms of the expected error types.

In that case, maybe `Other` should be called `UserInputValidated`?

Ooh, what about this: change them all to be named based on the error type rather than the operation or input.  That makes more sense next to the data that each variant contains, too.  Then it's not really an `OpKind` either...maybe it's:

```rust
enum UserErrorsExpected {
    NotFoundByResource(&dyn authz::ApiResourceError),
    NotFoundByLookup(ResourceType, LookupType),
    Conflict(ResourceType, &str),
    None,
}
```

> If we still think of this case in terms of input operation, it really is a catch-all for "something that isn't all the other cases"

I think that inference is incorrect and that's exactly what I'm worried about people thinking by calling it `Other`.  Today, `Other` _happens_ to correspond to all the cases that aren't covered by the other three.  But you could say that about any of the other three, too.  I'd argue it's not a catch-all, in that it would be wrong to assume that it's always a "safe" choice or that it would be correct for some new query that you're adding.  To be more concrete, if a developer adds some new query to the system, I'm afraid they're going to pick `Other` because their case isn't one of the first three.  What I want is for the developer to stop and ask whether some other 400-level error is possible and whether a new variant should be created.